lib/auth/join_bound_keypair.go

```diff
@@ -0,0 +1,483 @@
+}
+
+type createBoundKeypairValidator func(subject string, clusterName string, publicKey crypto.PublicKey) (boundKeypairValidator, error)
+
+// validateBoundKeypairTokenSpec performs some basic validation checks on a
+// bound_keypair-type join token.
+func validateBoundKeypairTokenSpec(spec *types.ProvisionTokenSpecV2BoundKeypair) error {
+	// Various constant checks, shared between creation and update. Many of
+	// these checks are temporary and will be removed alongside the experiment
+	// flag.
+	if !experiment.Enabled() {
+		return trace.BadParameter("bound keypair joining experiment is not enabled")
+	}
+
+	if spec.RotateOnNextRenewal {
+		return trace.NotImplemented("spec.bound_keypair.rotate_on_next_renewal is not yet implemented")
+	}
+
+	if spec.Onboarding.InitialJoinSecret != "" {
+		return trace.NotImplemented("spec.bound_keypair.initial_join_secret is not yet implemented")
+	}
+
+	if spec.Onboarding.InitialPublicKey == "" {
+		return trace.NotImplemented("spec.bound_keypair.initial_public_key is currently required")
+	}
+
+	if !spec.Joining.Unlimited {
+		return trace.NotImplemented("spec.bound_keypair.joining.unlimited cannot currently be `false`")
+	}
+
+	if !spec.Joining.Insecure {
+		return trace.NotImplemented("spec.bound_keypair.joining.insecure cannot currently be `false`")
+	}
+
+	return nil
+}
+
+func (a *Server) CreateBoundKeypairToken(ctx context.Context, token types.ProvisionToken) error {
+	if token.GetJoinMethod() != types.JoinMethodBoundKeypair {
+		return trace.BadParameter("must be called with a bound keypair token")
+	}
+
+	tokenV2, ok := token.(*types.ProvisionTokenV2)
+	if !ok {
+		return trace.BadParameter("%v join method requires ProvisionTokenV2", types.JoinMethodOracle)
+	}
+
+	spec := tokenV2.Spec.BoundKeypair
+	if spec == nil {
+		return trace.BadParameter("bound_keypair token requires non-nil spec.bound_keypair")
+	}
+
+	if err := validateBoundKeypairTokenSpec(spec); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// TODO: Is this wise? End users _shouldn't_ modify this, but this could
+	// interfere with cluster backup/restore. Options seem to be:
+	// - Let users create/update with status fields. They can break things, but
+	//   maybe that's okay. No backup/restore implications.
+	// - Ignore status fields during creation and update. Any set value will be
+	//   discarded here, and during update. This would still have consequences
+	//   during cluster restores, but wouldn't raise errors, and the status
+	//   field would otherwise be protected from easy tampering. Users might be
+	//   confused as no user-visible errors would be raised if they used
+	//   `tctl edit`.
+	// - Raise an error if status fields are changed. Worst restore
+	//   implications, but tampering won't be easy, and will have some UX.
+	if tokenV2.Status.BoundKeypair != nil {
+		return trace.BadParameter("cannot create a bound_keypair token with set status")
+	}
```

Comment:
Curious for thoughts on this. I think the simplest option is probably to just let users modify the resource freely, even if that has consequences.

Reply:
I think this is one of the reason a resource's "status" is actually a sub-resource (with different RBAC permissions) in Kubernetes.

Reply:
Hmm, that makes sense. I'll just leave a note and keep it simple - we'll tell users not to modify the status field, but there won't be any actual guardrails in place so we don't get in the way of restore attempts.